src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -587,7 +630,7 @@ internal abstract class EnumerableSorter<TElement>
         private int[] ComputeMap(TElement[] elements, int count)
         {
             ComputeKeys(elements, count);
-            int[] map = new int[count];
+            int[] map = ArrayPool<int>.Shared.Rent(count);
```

Comment:
There's potential for doing something with `ComputeKeys`, though it would likely want to check it was being used with a common key type. It also would need to guard against exceptions a lot more (the sort of case where it would be great if C# had `try…fault`, because you'd want to clean up within it if something went wrong, but return it otherwise). It's probably more complexity than it's worth, but perhaps worth experimenting.